### PR TITLE
support ACC adapter which has CU command queue

### PR DESCRIPTION
**Pull request by mamin506**

### CODE REVIEW

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 1]
```diff
@@ -789,13 +711,13 @@ configure(struct sched_cmd *cmd)
 
 	SCHED_DEBUG("Configuring scheduler\n");
 	exec->num_slots       = CQ_SIZE / cfg->slot_size;
+	write_lock(&zdev->attr_rwlock);
 	exec->num_cus         = cfg->num_cus;
+	write_unlock(&zdev->attr_rwlock);
```
Comment (xdavidz) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: There is a small window of race? 
we released lock here, but the exec->configured has not been set yet, the other threads may be able to check exec->configured == 0 and configure it again.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 2]
```diff
@@ -819,21 +741,31 @@ configure(struct sched_cmd *cmd)
...
 
-	exec->cu = vzalloc(sizeof(struct zocl_cu) * exec->num_cus);
+	exec->zcu = vzalloc(sizeof(struct zocl_cu_new) * exec->num_cus);
+	if (!exec->zcu) {
+		DRM_ERROR("Cound not allocate CU objects\n");
+		return -ENOMEM;
+	}
 
 	for (i = 0; i < exec->num_cus; i++) {
+		if (cfg->data[i] & ~ZOCL_KDS_MASK == ACCEL_ADAPTER) {
```
Comment (xdavidz) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: 19:45:21 | /opt/xrt/src/runtime_src/core/edge/drm/zocl/sched_exec.c:751:20: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
19:45:21 |    if (cfg->data[i] & ~ZOCL_KDS_MASK == ACCEL_ADAPTER) {

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 3]
```diff
@@ -819,21 +741,31 @@ configure(struct sched_cmd *cmd)
...
+			/* If the ACCEL adapter is used */
+			acc_cu = 1;
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: should't we clear acc_cu to zero someware inside the for loop?

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 4]
```diff
@@ -1243,22 +1131,34 @@ cu_done(struct sched_cmd *cmd)
...
 		unsigned int mask_idx = cu_mask_idx(cu_idx);
 		unsigned int pos = cu_idx_in_mask(cu_idx);
 
 		set_cmd_ext_timestamp(cmd, CU_DONE_TIME);
-		zdev->exec->cu_status[mask_idx] ^= 1<<pos;
+		/* if cu is busy, change the status to free */
+		if (zdev->exec->cu_status[mask_idx] & 1<<pos)
+			zdev->exec->cu_status[mask_idx] ^= 1<<pos;
+
+		/* This command is completed, refune credit and delete it */
```
Comment (larry9523) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: refund

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 1 above)
Comment (mamin506) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: Yes. Should include exec->configured.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 2 above)
Comment (mamin506) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: done

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 3 above)
Comment (mamin506) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: fixed

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 5]
```diff
@@ -1755,8 +1670,8 @@ configure_cu(struct sched_cmd *cmd, int cu_idx)
 	cmd->exectime = cu_idx_to_timeout(cmd->ddev, cmd->cu_idx);
 	cmd->check_timeout = cmd->exectime > 0 ? 1 : 0;
 
-	/* start CU at base + 0x0 */
-	iowrite32(0x1, virt_addr);
+	zocl_cu_start(cu);
```
Comment (xdavidz) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: I have some additional question, regrading the cu_done and cu_start.

zocl_cu_start iowrite32(0x1, cu_core->vaddr); 
list_add_tail(&cmd->rq_list, &cu->running_queue); 
 
ert_configure_cu doesn't add it to the running queue? I guess it changed the current behavior of handling cu? aka, set to 0x1 wait for 0x6? now we keep setting it?

Comment (mamin506) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: ert_configure_cu should do the same. Let me fix it.
We didn't keep setting it. Before submit the command to the CU. Scheduler need to get a credit from CU. The old behavior will like the credit of the CU is ways 1.
But if CU support Command Queue, the credit of the CU is > 1. So, before the credit burns out, the scheduler could always get a credit for the command and submit the command to a CU.

Comment (xdavidz) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: LGTM

**Comment by maxzhen** [comment]
retest this please

**Pull request merged**